File: app/src/main/java/space/pal/sig/view/news/NewsFragment.java
package space.pal.sig.view.news;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.lifecycle.ViewModelProvider;
import androidx.paging.PagedList;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import javax.inject.Inject;

import butterknife.BindView;
import space.pal.sig.R;
import space.pal.sig.Space;
import space.pal.sig.model.News;
import space.pal.sig.model.NewsSource;
import space.pal.sig.view.SpaceBaseFragment;

/**
 * SpaceNow
 * Created by Catalin on 7/17/2020
 **/
public class NewsFragment extends SpaceBaseFragment {

    @BindView(R.id.news_recycler)
    RecyclerView newsRecycler;
    @Inject
    NewsViewModelFactory factory;
    private NewsViewModel newsViewModel;
    private NewsAdapter newsAdapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_news,
                container, false);
        setHasOptionsMenu(true);
        setupBinding(this, view);
        Space.getApplicationComponent().inject(this);
        newsViewModel = new ViewModelProvider(this, factory)
                .get(NewsViewModel.class);
        newsRecycler.setLayoutManager(new LinearLayoutManager(getContext()));
        newsViewModel.getNewsList()
                .observe(getViewLifecycleOwner(), this::consumeNewsList);
        newsViewModel.getSelectedSource().observe(getViewLifecycleOwner(), this::onSelectSource);
        return view;
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.filter_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.filter) {
            String[] sources = newsViewModel.getNewsSources();
            AlertDialog dialog = new AlertDialog
                    .Builder(getParentActivity())
                    .setTitle(R.string.news_sources)
                    .setItems(sources, (d, w) -> {
                        newsViewModel.setSource(sources[w]);
                        d.dismiss();
                    })
                    .create();
            dialog.show();
        }
        return super.onOptionsItemSelected(item);
    }

    private void onSelectSource(NewsSource source) {
        newsViewModel.findNewsBySource(source);
        getParentActivity().setTitle(source.getText());
    }

    private void consumeNewsList(PagedList<News> newsPagedList) {
        newsAdapter = new NewsAdapter();
        newsRecycler.setAdapter(newsAdapter);
        newsAdapter.submitList(newsPagedList);
    }
}

File: app/src/main/java/space/pal/sig/view/news/NewsFragment.kt
package space.pal.sig.view.news

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import org.koin.android.ext.android.inject
import space.pal.sig.R
import space.pal.sig.databinding.FragmentNewsBinding
import space.pal.sig.model.entity.News
import space.pal.sig.view.BaseFragment
import space.pal.sig.view.WebActivity

/**
 * SpaceNow
 * Created by Catalin on 12/21/2020
 **/
class NewsFragment : BaseFragment(), NewsClickListener {

    private lateinit var binding: FragmentNewsBinding
    private val viewModel: NewsViewModel by inject()
    private val adapter = NewsAdapter(this)

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_news,
                container, false)
        viewModel.getNewsList()
                .observe(viewLifecycleOwner) { showNewsList(it) }
        binding.newsRecycler.layoutManager = LinearLayoutManager(requireContext())
        binding.newsRecycler.adapter = adapter
        return binding.root
    }

    override fun onClick(news: News) {
        val intent = Intent(requireContext(), WebActivity::class.java)
        intent.putExtra(WebActivity.URL, news.link)
        intent.putExtra(WebActivity.TITLE, news.title)
        startActivity(intent)
    }

    private fun showNewsList(newsList: MutableList<News>) {
        adapter.submit(newsList)
    }

}